homeassistant/components/romy/vacuum.py

```diff
@@ -0,0 +1,288 @@
+"""Support for Wi-Fi enabled ROMY vacuum cleaner robots.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/vacuum.romy/.
+"""
+
+from collections.abc import Mapping
+import json
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.vacuum import (
+    PLATFORM_SCHEMA,
+    VacuumEntity,
+    VacuumEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PASSWORD, CONF_PORT
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .utils import async_query
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = "mdi:robot-vacuum"
+PLATFORM = "romy"
+
+FAN_SPEED_NONE = "Default"
+FAN_SPEED_NORMAL = "Normal"
+FAN_SPEED_SILENT = "Silent"
+FAN_SPEED_INTENSIVE = "Intensive"
+FAN_SPEED_SUPER_SILENT = "Super_Silent"
+FAN_SPEED_HIGH = "High"
+FAN_SPEED_AUTO = "Auto"
+
+FAN_SPEEDS: list[str] = [
+    FAN_SPEED_NONE,
+    FAN_SPEED_NORMAL,
+    FAN_SPEED_SILENT,
+    FAN_SPEED_INTENSIVE,
+    FAN_SPEED_SUPER_SILENT,
+    FAN_SPEED_HIGH,
+    FAN_SPEED_AUTO,
+]
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT, default=8080): cv.port,
+        vol.Optional(CONF_PASSWORD): vol.All(str, vol.Length(8)),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+# Commonly supported features
+SUPPORT_ROMY_ROBOT = (
+    VacuumEntityFeature.BATTERY
+    | VacuumEntityFeature.PAUSE
+    | VacuumEntityFeature.RETURN_HOME
+    | VacuumEntityFeature.SEND_COMMAND
+    | VacuumEntityFeature.STATUS
+    | VacuumEntityFeature.STOP
+    | VacuumEntityFeature.TURN_OFF
+    | VacuumEntityFeature.TURN_ON
+    | VacuumEntityFeature.FAN_SPEED
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up ROMY vacuum with config entry."""
+    host = config_entry.data[CONF_HOST]
+    port = config_entry.data[CONF_PORT]
+    name = config_entry.data[CONF_NAME]
```

Comment:
Okay I'm going to make this comment here but it affects a large majority of the code.

You need to create a new pypi package to handle communicating with the robot.

https://developers.home-assistant.io/docs/creating_component_code_review/#4-communication-with-devicesservices


Then you should create a coordinator (There are other ways to do it - but this is the easiest imo.) You create that coordinator in async_setup_entry in init, and you set hass.data[domain] equal to it.


Once you do that, you can acess the coordinator anywhere you have access to hass, and you should make your vacuum entity a coordinatated entity, so you can use the coordinator to update, rather than logging in and accessing it directly through the robot entity.

Reply:
i managed somehow to over-read this, will start creating a pypi package
(sorry i clicked on request review by accident, can not undo it somehow)